app/src/main/java/org/simple/clinic/scanid/ScanSimpleIdScreenController.kt

```diff
@@ -17,11 +18,33 @@ class ScanSimpleIdScreenController @Inject constructor() : ObservableTransformer
   override fun apply(events: Observable<UiEvent>): ObservableSource<UiChange> {
     val replayedEvents = ReplayUntilScreenIsDestroyed(events)
         .compose(ReportAnalyticsEvents())
+        .compose(mergeWithScannedBpPassportCodes())
         .replay()
 
-    return replayedEvents
+    return sendScannedBpPassportCodes(replayedEvents)
+  }
+
+  private fun sendScannedBpPassportCodes(events: Observable<UiEvent>): Observable<UiChange> {
+    return events
+        .ofType<ScanSimpleIdScreenPassportCodeScanned.ValidPassportCode>()
+        .take(1)
+        .map { { ui: Ui -> ui.sendScannedPassportCode(it.bpPassportUuid) } }
+  }
+
+  private fun mergeWithScannedBpPassportCodes() = ObservableTransformer<UiEvent, UiEvent> { events ->
+
+    val scannedBpPassportCodes = events
         .ofType<ScanSimpleIdScreenQrCodeScanned>()
         .map { it.text }
-        .map { text -> { ui: Ui -> ui.showPatientSearchResults(text) } }
+        .map { scannedQrCode ->
+          try {
+            val bpPassportCode = UUID.fromString(scannedQrCode)
+            ScanSimpleIdScreenPassportCodeScanned.ValidPassportCode(bpPassportCode)
+          } catch (e: IllegalArgumentException) {
+            ScanSimpleIdScreenPassportCodeScanned.InvalidPassportCode
```

Comment:
Yes. The reason I added this is because I needed to split the stream of scanned QR codes into valid ones and invalid ones. We may not do anything with it right now, but it's good to have it if we need to provide feedback to users that the scanned QR code is invalid.

Reply:
This is not getting reported to Heap at the moment because the composition happens after reporting to Heap. I'll fix this.

Reply:
Always self-correcting :)